core/src/main/java/org/elasticsearch/index/snapshots/IndexShardSnapshotStatus.java

```diff
@@ -171,71 +144,105 @@ public synchronized void addProcessedFile(long size) {
+                                                 indexVersion, failure);
+    }
+
+    public static IndexShardSnapshotStatus newInitializing() {
+        return new IndexShardSnapshotStatus(Stage.INIT, 0L, 0L, 0, 0, 0, 0, 0, null);
+    }
+
+    public static IndexShardSnapshotStatus newFailed(final String failure) {
+        if (failure == null) {
+            throw new IllegalArgumentException("A failure description is required for a failed IndexShardSnapshotStatus");
+        }
+        return new IndexShardSnapshotStatus(Stage.FAILURE, 0L, 0L, 0, 0, 0, 0, 0, failure);
+    }
+
+    public static IndexShardSnapshotStatus newDone(final long startTime, final long totalTime, final int files, final long size) {
+        // The snapshot is done which means the number of processed files is the same as total
+        return new IndexShardSnapshotStatus(Stage.DONE, startTime, totalTime, files, files, size, size, 0, null);
+    }
+
+    /**
+     * Returns an immutable state of {@link IndexShardSnapshotStatus} at a given point in time.
+     */
+    public static class Copy {
+
+        private final Stage stage;
+        private final long startTime;
+        private final long totalTime;
+        private final int numberOfFiles;
+        private final int processedFiles;
+        private final long totalSize;
+        private final long processedSize;
+        private final long indexVersion;
+        private final String failure;
+
+        public Copy(final Stage stage, final long startTime, final long totalTime,
+                    final int numberOfFiles, final int processedFiles, final long totalSize, final long processedSize,
+                    final long indexVersion, final String failure) {
+            this.stage = stage;
+            this.startTime = startTime;
+            this.totalTime = totalTime;
+            this.numberOfFiles = numberOfFiles;
+            this.processedFiles = processedFiles;
+            this.totalSize = totalSize;
+            this.processedSize = processedSize;
+            this.indexVersion = indexVersion;
+            this.failure = failure;
+        }
+
+        public Stage getStage() {
+            return stage;
+        }
+
+        public long getStartTime() {
+            return startTime;
+        }
+
+        public long getTotalTime() {
+            return totalTime;
+        }
+
+        public int getNumberOfFiles() {
+            return numberOfFiles;
+        }
+
+        public int getProcessedFiles() {
+            return processedFiles;
+        }
+
+        public long getTotalSize() {
+            return totalSize;
+        }
+
+        public long getProcessedSize() {
+            return processedSize;
+        }
+
+        public long getIndexVersion() {
+            return indexVersion;
+        }
+
+        public String getFailure() {
+            return failure;
+        }
+
+        @Override
+        public String toString() {
+            return new StringBuilder()
```

Comment:
this looks a bit odd of a toString() implementation as it's very much targeted towards that one logging call site. Maybe change it to be more generic.